rerun_cpp/src/rerun/datatypes/tensor_buffer_ext.cpp

```diff
@@ -0,0 +1,112 @@
+
+            /// Construct a `TensorBuffer` from a `std::vector<int32_t>`.
+            inline TensorBuffer(std::vector<int32_t> i32) : TensorBuffer(TensorBuffer::i32(i32)) {}
+
+            /// Construct a `TensorBuffer` from a `std::vector<int64_t>`.
+            inline TensorBuffer(std::vector<int64_t> i64) : TensorBuffer(TensorBuffer::i64(i64)) {}
+
+            // TODO(#3380): f16 support
+            // /// Construct a `TensorBuffer` from a `std::vector<half>`.
+            // inline TensorBuffer(std::vector<half> f16) : TensorBuffer(TensorBuffer::f16(f16))
+            // {}
+
+            /// Construct a `TensorBuffer` from a `std::vector<float>`.
+            inline TensorBuffer(std::vector<float> f32) : TensorBuffer(TensorBuffer::f32(f32)) {}
+
+            /// Construct a `TensorBuffer` from a `std::vector<double>`.
+            inline TensorBuffer(std::vector<double> f64) : TensorBuffer(TensorBuffer::f64(f64)) {}
+
+            /// Number of elements in the buffer.
+            ///
+            /// You may NOT call this for JPEG buffers.
+            size_t num_elems() const;
+
+            // [CODEGEN COPY TO HEADER END]
+        };
+
+#undef TensorBuffer
+#else
+#define TensorBufferExt TensorBuffer
+#endif
+
+        /// Number of elements in the buffer.
+        size_t TensorBufferExt::num_elems() const {
+            switch (this->_tag) {
+                case detail::TensorBufferTag::NONE: {
+                    break; // Nothing to destroy
+                }
+                case detail::TensorBufferTag::U8: {
+                    return _data.u8.size();
+                }
+                case detail::TensorBufferTag::U16: {
+                    return _data.u16.size();
+                }
+                case detail::TensorBufferTag::U32: {
+                    return _data.u32.size();
+                }
+                case detail::TensorBufferTag::U64: {
+                    return _data.u64.size();
+                }
+                case detail::TensorBufferTag::I8: {
+                    return _data.i8.size();
+                }
+                case detail::TensorBufferTag::I16: {
+                    return _data.i16.size();
+                }
+                case detail::TensorBufferTag::I32: {
+                    return _data.i32.size();
+                }
+                case detail::TensorBufferTag::I64: {
+                    return _data.i64.size();
+                }
+                case detail::TensorBufferTag::F16: {
+                    return _data.f16.size();
+                }
+                case detail::TensorBufferTag::F32: {
+                    return _data.f32.size();
+                }
+                case detail::TensorBufferTag::F64: {
+                    return _data.f64.size();
+                }
+                case detail::TensorBufferTag::JPEG: {
+                    assert(false && "Can't ask for the number of elements in a JPEG");
```

Comment:
Putting assert here means that we get a hard crash on debug and undefined behavior on release.
If we're okay with crash on debug, we should _still_ `return 0` afterwards for release builds!